sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprNodeFactory.java

```diff
@@ -0,0 +1,309 @@
+    }
+
+    public DebugExpressionPair createFloatConstant(float value) {
+        LLVMExpressionNode node = contextReference.get().getNodeFactory().createSimpleConstantNoArray(value, PrimitiveType.FLOAT);
+        return DebugExpressionPair.create(node, DebugExprType.getFloatType(32));
+    }
+
+    public DebugExpressionPair createCastIfNecessary(DebugExpressionPair pair, DebugExprType type) {
+        checkError(pair, "cast");
+        if (pair.getType() == type) {
+            return pair;
+        }
+        LLVMExpressionNode node;
+        if (type.isUnsigned())
+            node = contextReference.get().getNodeFactory().createUnsignedCast(pair.getNode(), type.getLLVMRuntimeType());
+        else
+            node = contextReference.get().getNodeFactory().createSignedCast(pair.getNode(), type.getLLVMRuntimeType());
+        return DebugExpressionPair.create(node, type);
+    }
+
+    @SuppressWarnings("static-method")
+    public DebugExpressionPair createObjectMember(DebugExpressionPair receiver, String fieldName) {
+        Object baseMember = null;
+        String memberName = null;
+        if (receiver.getNode() instanceof DebugExprVarNode) {
+            baseMember = ((DebugExprVarNode) (receiver.getNode())).getMember();
+            memberName = ((DebugExprVarNode) (receiver.getNode())).getName();
+        } else if (receiver.getNode() instanceof DebugExprObjectMemberNode) {
+            baseMember = ((DebugExprObjectMemberNode) (receiver.getNode())).getMember();
+            memberName = ((DebugExprObjectMemberNode) (receiver.getNode())).getFieldName();
+        }
+        if (baseMember != null) {
+            DebugExprObjectMemberNode node = new DebugExprObjectMemberNode(fieldName, baseMember);
+            return DebugExpressionPair.create(node, node.getType());
+        }
+        throw DebugExprException.symbolNotFound(receiver.getNode(), memberName, null);
+    }
+
+    public DebugExpressionPair createFunctionCall(DebugExpressionPair functionPair, List<DebugExpressionPair> arguments) {
+        checkError(functionPair, "call(...)");
+        if (functionPair.getNode() instanceof DebugExprVarNode) {
+            DebugExprVarNode varNode = (DebugExprVarNode) functionPair.getNode();
+            LLVMExpressionNode node = varNode.createFunctionCall(arguments, globalScopes);
+            return DebugExpressionPair.create(node, varNode.getType());
+        }
+        throw DebugExprException.typeError(functionPair.getNode(), functionPair.getNode().toString());
+    }
+
+    @SuppressWarnings("static-method")
+    public DebugExpressionPair createArrayElement(DebugExpressionPair array, DebugExpressionPair index) {
+        Object baseMember = null;
+        DebugExprType baseType = null;
+        if (array.getNode() instanceof DebugExprVarNode) {
+            baseMember = ((DebugExprVarNode) (array.getNode())).getMember();
+            baseType = ((DebugExprVarNode) (array.getNode())).getType();
```

Comment:
Why is this not just using `array.getType()`? It should always be the same, and that way, you don't need the `if-instanceof` cascade at all (see comment below why I think you don't need the `baseMember`).